src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -3508,10 +3509,43 @@ public Object stringIndex(DynamicObject string, DynamicObject pattern, int start
             return nil();
         }
 
-        @Specialization(guards = { "isRubyString(pattern)", "!isBrokenCodeRange(pattern)", "!isSingleByteSearch(string, pattern)" })
+        @Specialization(guards = { "!isSingleByteOptimizable(string)", "isAsciiCompatible(string)", "isSingleBytePattern(pattern)" })
+        public Object stringIndexAsciiCompatible(DynamicObject string, DynamicObject pattern, int start,
+                @Cached("create()") RopeNodes.BytesNode bytesNode,
+                @Cached("create()") RopeNodes.EncodingLengthNode encodingLengthNode,
+                @Cached("createBinaryProfile()") ConditionProfile badStartProfile,
+                @Cached("create()") BranchProfile matchFoundProfile,
+                @Cached("create()") BranchProfile noMatchProfile) {
+
+            final Rope sourceRope = rope(string);
+            final int end = sourceRope.byteLength();
+            final byte[] sourceBytes = bytesNode.execute(sourceRope);
+            final byte searchByte = bytesNode.execute(rope(pattern))[0];
+
+            if (badStartProfile.profile(start < 0 || start >= end)) {
+                return nil();
+            }
+
+            for (int i = start; i < end; i++) {
+                if (sourceBytes[i] == searchByte) {
+                    matchFoundProfile.enter();
+                    return i;
+                }
+
+                if (sourceBytes[i] < 0) {
+                    final int characterLength = encodingLengthNode.executeLength(sourceRope.getEncoding(), sourceBytes, i, sourceBytes.length);
+                    i += characterLength - 1; // Subtract one to account for the loop increment on the next iteration.
```

Comment:
I think this would be nicer with a `while` and then the `else` branch increments `i` by one.
Also some `isASCII()` helper rather than the semi-obscure `< 0`.

Reply:
Maybe if `encodingLengthNode` is specialized/optimized enough we would not even need the `if` here?

Reply:
I tried a while loop and it was slower. I didn't investigate why, however.

Reply:
I have plans to replace `encodingLengthNode` with something faster taking advantage of the fact we know the string is `CR_VALID`. But that was out of scope for this.

Reply:
It sounds very weird a `while` loop would be slower here.
But maybe it degenerates in a non-counted-loop or so.

Reply:
> I have plans to replace `encodingLengthNode` with something faster taking advantage of the fact we know the string is `CR_VALID`. But that was out of scope for this.

That would only be valid for some encodings like UTF8 where US-ASCII characters are always distinct to surrogate/extending characters, no?

Anyway, I think it is valuable to optimize EncodingLengthNode to write simpler code for walking over characters.
But it doesn't need to be in this PR.

Reply:
This specialization is only for ASCII-compatible encodings anyway and that's mostly a function of the pattern string being ASCII-only. So all 7-bit values would have to be the same, UTF-8 or otherwise. The search string must be compatible with the source string. But, I think we could handle other cases nicer, for sure.